### PR TITLE
:seedling: Add passphrase validation functions - Closes #456

### LINKED ISSUE #456: Passphrase validation client side
Lisk-Nano right now makes use of some logic to trim whitespaces (front / end) for the mnemonic passphrases. Also checking for passphrase validation etc: https://github.com/LiskHQ/lisk-nano/blob/development/src/components/passphraseInput/index.js#L35-L71

We could have this logic in lisk-js to unify this among all the client using lisk-js

### PR DESCRIPTION
**Pull request by Tosch110**
Closes #456

### CODE REVIEW

File: src/passphrase/validation.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+export const countPassphraseWhitespaces = passphrase => {
+	const whitespaceMatches = passphrase.match(/\s/g);
+	return whitespaceMatches ? whitespaceMatches.length : 0;
+};
+
+export const countPassphraseWords = passphrase =>
+	passphrase.split(' ').filter(Boolean).length;
+
+export const countCapitalCharacters = passphrase => {
```
Comment (willclarktech) [inline src/passphrase/validation.js]: Prefer `Uppercase` to `Capital`

File: src/passphrase/validation.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	const capitalCharacterMatches = passphrase.match(/[A-Z]/g);
+	return capitalCharacterMatches ? capitalCharacterMatches.length : 0;
+};
+
+export const validatePassphrase = passphrase => {
+	const expectedWords = 12;
+	const expectedWhitespaces = 11;
+	const expectedCapitalLetters = 0;
```
Comment (willclarktech) [inline src/passphrase/validation.js]: `UppercaseCharacters`

File: src/passphrase/validation.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		throw new Error(
+			`Passphrase contains ${whiteSpacesInPassphrase} whitespaces instead of expected ${expectedWhitespaces}. Please check the passphrase.`,
+		);
+	}
+	if (capitalCharacterInPassphrase !== expectedCapitalLetters) {
+		throw new Error(
+			`Passphrase contains ${capitalCharacterInPassphrase} capital character instead of expected ${expectedCapitalLetters}. Please check the passphrase.`,
+		);
+	}
+	if (!Mnemonic.validateMnemonic(passphrase)) {
```
Comment (willclarktech) [inline src/passphrase/validation.js]: This check covers some of the cases above. We're adding them to have more specific messages, correct?

Comment (willclarktech) [inline src/passphrase/validation.js]: Does this cover all the language requirements we have?

File: test/passphrase/validation.js [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+ */
+import {
+	countCapitalCharacters,
+	countPassphraseWhitespaces,
+	countPassphraseWords,
+	validatePassphrase,
+} from '../../src/passphrase/validation';
+
+describe('passphrase validation', () => {
+	describe('countPassphraseWhitespaces', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Do we really need to export and test these simple helper functions separately to the main validation function? It seemed like they were extracted for readability rather than because they abstract away some serious logic.

File: test/passphrase/validation.js [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+				countPassphraseWhitespaces(passphrase).should.be.equal(
+					expectedAmountOfWhitespaces,
+				);
+			});
+		});
+	});
+
+	describe('countPassphraseWords', () => {
+		describe('given a valid passphrase', () => {
+			const expextedAmountOfWords = 12;
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Spelling

File: test/passphrase/validation.js [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+		describe('given a valid passphrase', () => {
+			const passphrase =
+				'model actor shallow eight glue upper seat lobster reason label enlist bridge';
+
+			it('should return true', () => {
+				validatePassphrase(passphrase).should.be.true();
+			});
+		});
+
+		describe('given a passphrase with extra words', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: `too many`

File: test/passphrase/validation.js [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+				'model actor shallow eight glue upper seat lobster reason label enlist bridge actor';
+			const errorMessage =
+				'Passphrase contains 13 words instead of expected 12. Please check the passphrase.';
+
+			it('should throw the error', () => {
+				validatePassphrase.bind(null, passphrase).should.throw(errorMessage);
+			});
+		});
+
+		describe('given a passphrase with less words', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: `too few`

File: test/passphrase/validation.js [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+				'model actor shallow eight glue upper seat lobster reason label enlist';
+			const errorMessage =
+				'Passphrase contains 11 words instead of expected 12. Please check the passphrase.';
+
+			it('should throw the error', () => {
+				validatePassphrase.bind(null, passphrase).should.throw(errorMessage);
+			});
+		});
+
+		describe('given a passphrase with one more whitespace in the beginning', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: `extra` or `an extra`

File: test/passphrase/validation.js [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+				' model actor shallow eight glue upper seat lobster reason label enlist bridge';
+			const errorMessage =
+				'Passphrase contains 12 whitespaces instead of expected 11. Please check the passphrase.';
+
+			it('should throw the error', () => {
+				validatePassphrase.bind(null, passphrase).should.throw(errorMessage);
+			});
+		});
+
+		describe('given a passphrase with one more whitespace in the end', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: As above

File: test/passphrase/validation.js [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+				'model actor shallow eight glue upper seat lobster reason label enlist bridge ';
+			const errorMessage =
+				'Passphrase contains 12 whitespaces instead of expected 11. Please check the passphrase.';
+
+			it('should throw the error', () => {
+				validatePassphrase.bind(null, passphrase).should.throw(errorMessage);
+			});
+		});
+
+		describe('given a passphrase with more whitespaces in between', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: `extra whitespaces in the middle` or `in between words`

**Comment by willclarktech** [comment]
In addition to the above review, after https://github.com/LiskHQ/lisk-js/pull/506, all `it` blocks should return.

File: src/passphrase/validation.js (hunk 3 above)
Comment (Tosch110) [reply src/passphrase/validation.js]: Yes.

Comment (Tosch110) [reply src/passphrase/validation.js]: Good catch/question. Will go through it!

File: test/passphrase/validation.js (hunk 4 above)
Comment (Tosch110) [reply test/passphrase/validation.js]: I thought better have more than less tests regarding this logic. Since it can be quite sensitive.

Comment (willclarktech) [reply test/passphrase/validation.js]: As per discussion, different clients may have different requirements for passphrase validation and these functions are exported for their convenience (and hence require unit tests).

File: src/passphrase/validation.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import Mnemonic from 'bip39';
+
+export const countPassphraseWhitespaces = passphrase => {
```
[the remark this answers is not in the document]
Comment (Tosch110) [reply src/passphrase/validation.js]: Yes, I will be adding more tests to cover other whitespace charakter

File: src/passphrase/validation.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+	const whitespaceMatches = passphrase.match(/\s/g);
+	return whitespaceMatches ? whitespaceMatches.length : 0;
+};
+
+export const countPassphraseWords = passphrase =>
+	passphrase.split(' ').filter(Boolean).length;
+
+export const countUppercaseCharacters = passphrase => {
+	const uppercaseCharacterMatches = passphrase.match(/[A-Z]/g);
+	return uppercaseCharacterMatches ? uppercaseCharacterMatches.length : 0;
```
Comment (Reshetnyak) [inline src/passphrase/validation.js]: I'm not aware of your guidelines, but this function could be shorter
```javascript
// option 1
const uppercaseCharacterMatches = passphrase.match(/[A-Z]/g) || [];
return uppercaseCharacterMatches.length;
// option 2, just in one line. But not sure about readability.
export const countUppercaseCharacters = passphrase => (passphrase.match(/[A-Z]/g) || []).length;
```

The same could be applied to the function `countPassphraseWhitespaces `

Comment (willclarktech) [reply src/passphrase/validation.js]: Hi @Reshetnyak, thanks for the input, I actually almost suggested option 2 in my review (I always like a direct return from an arrow function 🙂 ). In this case I think readability/intelligibility is clearest with the longer option currently in place.

**Comment by Tosch110** [comment]
@slaweet 

Here is a proposal for the error messages. I have done a lot that you requested but not all. Still, we can discuss these things before a merge.

So now the validate function returns instead of throwing errors. It has an error code, with a message including the details but also expected state vs. is state. 
I did not include at which places those (uppercase, extra whitespace) occur because I could not think of a good method to put this understandable into the error message.

File: test/passphrase/validation.js [hunk 13]
```diff
@@ -229,12 +229,13 @@ describe('passphrase validation', () => {
 
 	describe('countUppercaseCharacters', () => {
 		describe('given a passphrase without uppercase character', () => {
+			const expectedAmountUppercaseCharacter = 0;
 			const passphrase =
 				'model actor shallow eight glue upper seat lobster reason label enlist bridge';
 
-			it('should return 0', () => {
+			it('should return the amount of uppercase character', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: `the number of uppercase characters`

File: test/passphrase/validation.js [hunk 14]
```diff
@@ -263,103 +264,211 @@ describe('passphrase validation', () => {
 	});
 
 	describe('validatePassphrase', () => {
+		const passphraseIsValid = [];
```
Comment (willclarktech) [inline test/passphrase/validation.js]: This is a confusing name now.

File: test/passphrase/validation.js [hunk 15]
```diff
@@ -263,103 +264,211 @@ describe('passphrase validation', () => {
...
+
 		describe('given a valid passphrase', () => {
 			const passphrase =
 				'model actor shallow eight glue upper seat lobster reason label enlist bridge';
 
-			it('should return true', () => {
-				return validatePassphrase(passphrase).should.be.true();
+			it('should return an array without the errors', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: `with no errors`: `without the errors` sounds like there are errors, they just shouldn't be returned.

File: test/passphrase/validation.js [hunk 16]
```diff
@@ -263,103 +264,211 @@ describe('passphrase validation', () => {
...
 			const passphrase =
 				'model actor shallow eight glue upper seat lobster reason label enlist bridge actor';
-			const errorMessage =
-				'Passphrase contains 13 words instead of expected 12. Please check the passphrase.';
 
-			it('should throw the error', () => {
-				return validatePassphrase
-					.bind(null, passphrase)
-					.should.throw(errorMessage);
+			const passphraseTooManyWordsError = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Should be plural. (Also below.)

File: test/passphrase/validation.js [hunk 17]
```diff
@@ -368,22 +477,31 @@ describe('passphrase validation', () => {
 			const passphrase =
 				'model actor shallow eight glue upper seat lobster reason label enlist bridge';
 
-			it('should return true', () => {
-				return validatePassphrase(passphrase, wordlist).should.be.true();
+			it('should return the array without the error', () => {
```
Comment (willclarktech) [inline test/passphrase/validation.js]: As above

File: src/passphrase/validation.js [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+			code: 'INVALID_MNEMONIC',
+			message:
+				'Passphrase is not a valid mnemonic passphrase. Please check the passphrase.',
+			expected: true,
+			actual: false,
+		};
+		errors.push(validationError);
+	}
+
+	return errors;
```
Comment (willclarktech) [inline src/passphrase/validation.js]: I still find it confusing that the return value for a success case is an empty array of errors. This seems really odd:
```
if (!validatePassphrase(passphrase).length) {
    // do something with a valid passphrase
}
```
Perhaps we need a different name, like `getPassphraseValidationErrors`?

File: test/passphrase/validation.js [hunk 19]
```diff
@@ -0,0 +1,508 @@
...
+				return validatePassphrase(passphrase).should.be.eql(
+					passphraseTooManyWordsErrors,
+				);
+			});
+		});
+
+		describe('given a passphrase with too few words', () => {
+			const passphrase =
+				'model actor shallow eight glue upper seat lobster reason label enlist';
+			const passphraseTooFewWordsError = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Plural

File: test/passphrase/validation.js [hunk 20]
```diff
@@ -0,0 +1,508 @@
...
+				return validatePassphrase(passphrase).should.be.eql(
+					passphraseTooFewWordsError,
+				);
+			});
+		});
+
+		describe('given a passphrase with an extra whitespace in the beginning', () => {
+			const passphrase =
+				' model actor shallow eight glue upper seat lobster reason label enlist bridge';
+			const passphraseTooManyWhitespacesError = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Plural

File: test/passphrase/validation.js [hunk 21]
```diff
@@ -0,0 +1,508 @@
...
+				return validatePassphrase(passphrase).should.be.eql(
+					passphraseTooManyWhitespacesError,
+				);
+			});
+		});
+
+		describe('given a passphrase with an extra whitespace in the end', () => {
+			const passphrase =
+				'model actor shallow eight glue upper seat lobster reason label enlist bridge ';
+			const passphraseTooManyWhitespacesError = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Plural

File: test/passphrase/validation.js [hunk 22]
```diff
@@ -0,0 +1,508 @@
...
+				return validatePassphrase(passphrase).should.be.eql(
+					passphraseTooManyWhitespacesError,
+				);
+			});
+		});
+
+		describe('given a passphrase with too many whitespaces in between words', () => {
+			const passphrase =
+				'model actor shallow eight glue  upper seat  lobster reason label enlist bridge';
+			const passphraseTooManyWhitespacesError = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Plural

File: test/passphrase/validation.js [hunk 23]
```diff
@@ -0,0 +1,508 @@
...
+				return validatePassphrase(passphrase).should.be.eql(
+					passphraseTooManyWhitespacesError,
+				);
+			});
+		});
+
+		describe('given a passphrase with uppercase characters', () => {
+			const passphrase =
+				'modEl actor shallow eight glue upper sEat lobster reaSon label enlist bridge';
+			const passphraseWithUppercaseCharacterError = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Plural

File: test/passphrase/validation.js [hunk 24]
```diff
@@ -0,0 +1,508 @@
...
+				return validatePassphrase(passphrase).should.be.eql(
+					passphraseWithUppercaseCharacterError,
+				);
+			});
+		});
+
+		describe('given a passphrase that is an invalid mnemonic passphrase', () => {
+			const passphrase =
+				'model actor shallow eight glue upper seat lobster reason label engage bridge';
+			const passphraseInvalidMnemonic = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: `Error`

File: test/passphrase/validation.js [hunk 25]
```diff
@@ -0,0 +1,508 @@
...
+					emptyErrorArray,
+				);
+			});
+		});
+
+		describe('given a passphrase that uses a different wordlist for the passphrase', () => {
+			const wordlist = Mnemonic.wordlists.spanish;
+			const passphrase =
+				'model actor shallow eight glue upper seat lobster reason label enlist bridge';
+			const passphraseInvalidMnemonic = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: `Error`

File: test/passphrase/validation.js [hunk 26]
```diff
@@ -0,0 +1,589 @@
...
+				return getPassphraseValidationErrors(passphrase).should.be.eql(
+					passphraseTooFewWordsErrors,
+				);
+			});
+		});
+
+		describe('given a passphrase with an extra whitespace in the beginning', () => {
+			const passphrase =
+				' model actor shallow eight glue upper seat lobster reason label enlist bridge';
+			const passphraseTooManyWhitespacesErrorsss = [
```
Comment (willclarktech) [inline test/passphrase/validation.js]: Prefer English over Parseltongue.
